design/cluster-api-provider-metal3/dhcpless_ipam.md

```diff
@@ -0,0 +1,397 @@
+reflected as ignored via the Metal3PreprovisioningTemplate status.
+
+The preprovisioningNetworkDataName is used by default for networkData in the baremetal-operator, so in this configuration it's not
+strictly necessary to specify networkData via Metal3DataTemplate, however we'll want to delete the IPClaim after preprovisioning 
+in the decoupled flow below so it seems likely we'll want to behave consistently and rely on the IP Reuse functionality if a
+consistent IP is required between pre-provisioning and provisioning phases.
+
+#### Preprovisioning Decoupled IPPool
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool-1
+spec:
+  clusterName: cluster
+
+---
+
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: preprovisioning-pool
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3PreprovisioningDataTemplate
+metadata:
+  name: preprov-data-template
+spec:
+  preprovisioningNetworkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: preprovisioning-pool
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3PreprovisioningTemplate
+metadata:
+  name: preprov-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: preprov-data-template
+      hostSelector:
+        matchLabels:
+          pre-provisioning: foo
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3DataTemplate
+metadata:
+  name: data-template
+spec:
+  clusterName: cluster
+  networkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: data-template
+      hostSelector:
+        matchLabels:
+          cluster-role: control-plane
+
+```
+
+In this flow we have `preprovisioning-pool` which is not associated with any cluster, this is used to provide an IPAddress during
+the pre-provisioning phase as described above.  To reduce the required size of the pool, the IPClaim will be deleted after the
+preprovisioning phase is completed, e.g the BMH resource becomes available.
+
+In the provisioning phase another pool, associated with a cluster is used to template networkData as in the existing process.
+
+#### Assumptions and Open Questions
+
+TODO
+
+### Inspection on initial registration
+
+On initial registration of a host, inspection is triggered immediately but this process cannot complete without preprovisioning network configuration in a DHCP-less environment (because the IPA ramdisk can't connect back to the Ironic API).
+
+To resolve this we can add a new BareMetalHost API `PreprovisioningNetworkDataRequired` which defaults to false, but when set to true will describe that the host cannot move from Registering -> Inspecting until `preprovisioningNetworkDataName` has been set.
```

Comment:
Possible alternative:
You create the BMH with a preprovisioningNetworkDataName pointing to an empty Secret. BMO refuses to start inspecting until the Secret contains some data. CAPM3 then fills in the Secret using the template/IPAM.

WDYT?

Reply:
We have an inspect annotations already, we can add a new value to it ("delayed"?).

Reply:
@dtantsur that would solve the problem of waiting to start inspection, but not the problem of how to associate the template used to create the preprovisioningNetworkData with the BMH, so IIUC from the threads elsewhere with @pierrecregut @Rozzii and @zaneb is to add a new annotation similar to:

```
baremetalhost.metal3.io/preprovisioningnetworkdata-reference: metal3preprovisioningdatatemplates.infrastructure.cluster.x-k8s.io/a-template
```

This is a little verbose, but it enables us to tell BMO that *something* is expected to configure a preprovisioningNetworkData secret, so we can wait until that happens, and whatever generates the secret can read the annotation value (in this proposal CAPM3, but as we discussed yesterday that could be made pluggable in future, so this approach avoids coupling the BMH API to a specific template resource type)

Reply:
I reworked the proposal to use an annotation as described, but included the suggestion from @zaneb and the `PreprovisioningNetworkDataRequired` option in the alternatives section